numba/targets/base.py

```diff
@@ -913,18 +916,26 @@ def make_constant_array(self, builder, typ, ary):
         Create an array structure reifying the given constant array.
         A low-level contiguous array constant is created in the LLVM IR.
         """
-        assert typ.layout == 'C'                # assumed in typeinfer.py
         datatype = self.get_data_type(typ.dtype)
-
-        # Handle data: reify the flattened array in "C" order as a
-        # global array of bytes.
-        flat = ary.flatten()
-        # Note: we use `bytearray(flat.data)` instead of `bytearray(flat)` to
-        #       workaround issue #1850 which is due to numpy issue #3147
-        consts = Constant.array(Type.int(8), bytearray(flat.data))
-        data = cgutils.global_constant(builder, ".const.array.data", consts)
-        # Ensure correct data alignment (issue #1933)
-        data.align = self.get_abi_alignment(datatype)
+        # don't freeze ary of non-C contig or bigger than 1MB
+        size_limit = 10**6
+        if (self.allow_dynamic_globals and
+                typ.layout != 'C' or ary.nbytes > size_limit):
```

Comment:
Can a non-C array happen here? The old comment ("assumed in typeinfer.py") says no, and this PR doesn't modify type inference.